File: python/src/gencmu/_markdown.py
"""The two parts of reading Markdown that are not grammars: the fenced
``ebnf`` blocks of a grammar document (engine §8) and the processing
instructions of a pipeline document (design, "Pipelines")."""

from __future__ import annotations

import posixpath
import re
from dataclasses import dataclass, field

from ._errors import GencmuError

_FENCE = re.compile(r"^ {0,3}(`{3,}|~{3,})(.*)$")
_LINE_BREAK = re.compile(r"\r\n|\r|\n")


def split_lines(text: str) -> list[str]:
    """The lines of a text, split at ``\\n``, ``\\r\\n`` and ``\\r``."""
    return _LINE_BREAK.split(text)


def line_column(text: str, index: int) -> tuple[int, int]:
    """The 1-based line and column of a code point index."""
    line = 1
    start = 0
    for match in _LINE_BREAK.finditer(text, 0, index):
        if match.end() > index:
            break
        line += 1
        start = match.end()
    return line, index - start + 1


@dataclass
class GrammarText:
    """The grammar text of a document: its ``ebnf`` blocks joined with a
    newline between them, and each character's line and column in the
    document, with one more entry for the end of the text."""

    text: str
    positions: list[tuple[int, int]]

    def position(self, index: int) -> tuple[int, int]:
        if not self.positions:
            return (1, 1)
        return self.positions[min(max(index, 0), len(self.positions) - 1)]


def ebnf_text(document: str) -> GrammarText:
    """Find the fenced ``ebnf`` blocks of a Markdown document (engine §8)."""
    lines = split_lines(document)
    blocks: list[list[tuple[int, str]]] = []
    index = 0
    while index < len(lines):
        match = _FENCE.match(lines[index])
        if match is None:
            index += 1
            continue
        fence, info = match.group(1), match.group(2)
        if fence[0] == "`" and "`" in info:
            index += 1
            continue
        closing = re.compile(r"^ {0,3}" + re.escape(fence[0]) + "{" + str(len(fence)) + r",}[ \t]*$")
        body: list[tuple[int, str]] = []
        opening = index + 1
        index += 1
        while index < len(lines) and closing.match(lines[index]) is None:
            body.append((index + 1, lines[index]))
            index += 1
        if index >= len(lines):
            # Were a block that is never closed to run to the end, a missing
            # fence would silently make the rest of the document grammar, or
            # hide it.
            raise GencmuError("a fenced code block is never closed", line=opening, column=1)
        index += 1
        if info.strip() == "ebnf":
            blocks.append(body)
    chars: list[str] = []
    positions: list[tuple[int, int]] = []
    last = (1, 1)
    for number, block in enumerate(blocks):
        if number > 0:
            chars.append("\n")
            positions.append(last)
        for row, (line_number, line) in enumerate(block):
            if row > 0:
                chars.append("\n")
                positions.append(last)
            for column, char in enumerate(line, 1):
                chars.append(char)
                positions.append((line_number, column))
            last = (line_number, len(line) + 1)
    positions.append(last)
    return GrammarText("".join(chars), positions)


_INSTRUCTION = re.compile(r"<\?(stage|grammar|features)(?=[\s?])([^?]*)\?>")
_LINK = re.compile(r"\[[^\]]*\]\(([^\s()\\]*)\)")
_FEATURE = re.compile(r"[A-Za-z][A-Za-z0-9-]*")
_STAGE_NAME = re.compile(r"[^\s<>?]+")


@dataclass
class PipelineStage:
    name: str
    documents: list[str] = field(default_factory=list)


@dataclass
class Pipeline:
    stages: list[PipelineStage]
    features: frozenset[str]


def resolve(base: str, target: str) -> str:
    """A link target resolved against the path of the document it is in."""
    joined = posixpath.join(posixpath.dirname(base), target) if not target.startswith("/") else target
    normalized = posixpath.normpath(joined)
    if normalized.startswith("/"):
        normalized = normalized.lstrip("/")
    return normalized


def read_pipeline(text: str, path: str) -> Pipeline:
    """Read a pipeline document's stages, documents and features."""
    stages: list[PipelineStage] = []
    features: set[str] = set()
    names: set[str] = set()
    for number, line in enumerate(split_lines(text), 1):
        found = list(_INSTRUCTION.finditer(line))
        if not found:
            continue

        def fail(message: str) -> GencmuError:
            return GencmuError(message, document=path, line=number, column=found[0].start() + 1)

        if len(found) > 1:
            raise fail("a line holds at most one processing instruction")
        match = found[0]
        if line[match.end():].strip():
            continue
        kind, argument = match.group(1), match.group(2).strip()
        if kind == "stage":
            if not line.lstrip().startswith("#"):
                raise fail("<?stage?> must end a heading")
            if not _STAGE_NAME.fullmatch(argument):
                raise fail(f"a malformed stage name: {argument!r}")
            if argument in names:
                raise fail(f"two stages are named {argument}")
            names.add(argument)
            stages.append(PipelineStage(argument))
        elif kind == "grammar":
            if argument:
                raise fail("<?grammar?> takes no argument")
            if not stages:
                raise fail("<?grammar?> before any <?stage?>")
            link = _LINK.search(line[: match.start()])
            if link is None or not link.group(1):
                raise fail("a <?grammar?> line without a link [text](path)")
            stages[-1].documents.append(resolve(path, link.group(1)))
        else:
            words = argument.split()
            if not words:
                raise fail("<?features?> names no feature")
            for word in words:
                if not _FEATURE.fullmatch(word):
                    raise fail(f"a malformed feature name: {word!r}")
            features.update(words)
    if not stages:
        raise GencmuError("the pipeline has no <?stage?>", document=path)
    return Pipeline(stages, frozenset(features))

File: go/markdown.go
package gencmu

import (
	"fmt"
	"path"
	"regexp"
	"strings"
	"unicode/utf8"
)

// Error is a dialect that cannot be loaded: a document is missing, does not
// parse as the notation, or does not stitch into a valid grammar. Document,
// Line and Column say where, when known; Line and Column count from 1, in
// code points.
type Error struct {
	Kind     string // always "grammar" for a load error
	Document string
	Line     int
	Column   int
	Stage    string
	Message  string
}

func (e *Error) Error() string {
	var b strings.Builder
	if e.Document != "" {
		b.WriteString(e.Document)
		if e.Line > 0 {
			fmt.Fprintf(&b, ":%d:%d", e.Line, e.Column)
		}
		b.WriteString(": ")
	}
	if e.Stage != "" {
		fmt.Fprintf(&b, "stage %s: ", e.Stage)
	}
	b.WriteString(e.Message)
	return b.String()
}

func grammarError(document string, at [2]int, format string, args ...any) *Error {
	return &Error{Kind: "grammar", Document: document, Line: at[0], Column: at[1], Message: fmt.Sprintf(format, args...)}
}

// splitLines splits at \n, \r\n and \r, the line ends of every position in
// gencmu.
func splitLines(s string) []string {
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '\n':
			lines = append(lines, s[start:i])
			start = i + 1
		case '\r':
			lines = append(lines, s[start:i])
			if i+1 < len(s) && s[i+1] == '\n' {
				i++
			}
			start = i + 1
		}
	}
	return append(lines, s[start:])
}

// grammarText is the text of a document's ebnf blocks (engine §8), with the
// document position of each of its code points and of its end.
type grammarText struct {
	text []rune
	pos  [][2]int // len(text)+1 entries: line and column, from 1
}

func (g *grammarText) at(offset int) [2]int {
	if offset < 0 {
		offset = 0
	}
	if offset >= len(g.pos) {
		offset = len(g.pos) - 1
	}
	return g.pos[offset]
}

var fenceOpen = regexp.MustCompile("^ {0,3}(`{3,}|~{3,})(.*)$")

// extractEBNF finds the fenced code blocks whose info string is ebnf and joins
// their contents with a newline between blocks. A block's lines are joined
// with a newline too, so every line ending reads as \n.
func extractEBNF(doc string) *grammarText {
	g := &grammarText{}
	lines := splitLines(doc)
	var fence string // the open fence, or "" outside a block
	isEBNF, blocks, linesInBlock := false, 0, 0
	lastEnd := [2]int{1, 1}
	newline := func() {
		g.text = append(g.text, '\n')
		g.pos = append(g.pos, lastEnd)
	}
	for n, line := range lines {
		if fence == "" {
			m := fenceOpen.FindStringSubmatch(line)
			if m == nil {
				continue
			}
			info := strings.TrimSpace(m[2])
			if m[1][0] == '`' && strings.Contains(info, "`") {
				continue
			}
			fence = m[1]
			isEBNF = info == "ebnf"
			if isEBNF {
				if blocks > 0 {
					newline()
				}
				blocks++
				linesInBlock = 0
			}
			continue
		}
		if isClosingFence(line, fence) {
			fence = ""
			continue
		}
		if !isEBNF {
			continue
		}
		if linesInBlock > 0 {
			newline()
		}
		linesInBlock++
		col := 1
		for _, c := range line {
			g.text = append(g.text, c)
			g.pos = append(g.pos, [2]int{n + 1, col})
			col++
		}
		lastEnd = [2]int{n + 1, col}
	}
	g.pos = append(g.pos, lastEnd)
	return g
}

// isClosingFence: a fence of the same character at least as long, indented
// up to three spaces, followed by nothing but spaces.
func isClosingFence(line, fence string) bool {
	t := strings.TrimLeft(line, " ")
	if len(line)-len(t) > 3 {
		return false
	}
	n := 0
	for n < len(t) && t[n] == fence[0] {
		n++
	}
	return n >= len(fence) && strings.TrimRight(t[n:], " \t") == ""
}

// A pipeline document (design, "Pipelines"): stages, the documents stitched
// into each, and the features the dialect enables.
type pipelineStage struct {
	name      string
	documents []string // resolved paths
	at        [2]int
}

type pipeline struct {
	stages   []pipelineStage
	features []string
}

var (
	instruction = regexp.MustCompile(`<\?([A-Za-z][A-Za-z0-9-]*)((?:\s[^?]*)?)\?>`)
	heading     = regexp.MustCompile(`^ {0,3}#{1,6}(\s|$)`)
	linkStart   = regexp.MustCompile(`\[[^\]]*\]\(`)
	featureName = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9-]*$`)
)

func runeColumn(line string, byteOffset int) int {
	return utf8.RuneCountInString(line[:byteOffset]) + 1
}

// readPipeline reads a pipeline document's processing instructions, one per
// line at most, each at the end of its line.
func readPipeline(text, docPath string) (*pipeline, *Error) {
	p := &pipeline{}
	seen := map[string]bool{}
	features := map[string]bool{}
	for n, line := range splitLines(text) {
		found := instruction.FindAllStringSubmatchIndex(line, -1)
		if len(found) == 0 {
			continue
		}
		at := [2]int{n + 1, runeColumn(line, found[0][0])}
		if len(found) > 1 {
			second := [2]int{n + 1, runeColumn(line, found[1][0])}
			return nil, grammarError(docPath, second, "a line of a pipeline holds at most one processing instruction")
		}
		m := found[0]
		if strings.TrimSpace(line[m[1]:]) != "" {
			continue // not at the end of the line: prose
		}
		name, args := line[m[2]:m[3]], strings.TrimSpace(line[m[4]:m[5]])
		switch name {
		case "stage":
			if !heading.MatchString(line) {
				return nil, grammarError(docPath, at, "<?stage?> must end a heading")
			}
			if args == "" || strings.ContainsAny(args, " \t") {
				return nil, grammarError(docPath, at, "<?stage?> needs one stage name")
			}
			if seen[args] {
				return nil, grammarError(docPath, at, "two stages are named %s", args)
			}
			seen[args] = true
			p.stages = append(p.stages, pipelineStage{name: args, at: at})
		case "grammar":
			if args != "" {
				return nil, grammarError(docPath, at, "<?grammar?> takes no arguments")
			}
			if len(p.stages) == 0 {
				return nil, grammarError(docPath, at, "a <?grammar?> line comes before any stage")
			}
			l := linkStart.FindStringIndex(line[:m[0]])
			if l == nil {
				return nil, grammarError(docPath, at, "a <?grammar?> line needs a link [text](path)")
			}
			rest := line[l[1]:m[0]]
			end := strings.IndexByte(rest, ')')
			target := ""
			if end >= 0 {
				target = rest[:end]
			}
			if end < 0 || target == "" || strings.ContainsAny(target, " \t()\\") {
				return nil, grammarError(docPath, [2]int{n + 1, runeColumn(line, l[0])}, "a <?grammar?> link must be [text](path) with no spaces, parentheses or backslashes in the path")
			}
			s := &p.stages[len(p.stages)-1]
			s.documents = append(s.documents, resolvePath(docPath, target))
		case "features":
			names := strings.Fields(args)
			if len(names) == 0 {
				return nil, grammarError(docPath, at, "<?features?> names no feature")
			}
			for _, f := range names {
				if !featureName.MatchString(f) {
					return nil, grammarError(docPath, at, "%q is not a feature name", f)
				}
				if !features[f] {
					features[f] = true
					p.features = append(p.features, f)
				}
			}
		default:
			// Another tool's instruction: not the engine's to read.
		}
	}
	if len(p.stages) == 0 {
		return nil, grammarError(docPath, [2]int{}, "the pipeline has no <?stage?>")
	}
	return p, nil
}

// resolvePath resolves a link target against the document holding it, with
// . and .. normalized.
func resolvePath(from, target string) string {
	return path.Clean(path.Join(path.Dir(from), target))
}
